calyx/src/passes/top_down_compile_control.rs

```diff
@@ -367,118 +424,168 @@ fn calculate_states_recur(
-            cond,
-            tbranch,
-            fbranch,
-            ..
-        }) => {
-            if cond.is_some() {
-                return Err(Error::malformed_structure(format!("{}: Found group `{}` in with position of if. This should have compiled away.", TopDownCompileControl::name(), cond.as_ref().unwrap().borrow().name())));
-            }
-            let port_guard: ir::Guard = Rc::clone(port).into();
-            // Previous states transitioning into true branch need the conditional
-            // to be true.
-            let tru_transitions = preds.clone().into_iter().map(|(s, g)| (s, g & port_guard.clone())).collect();
-            let tru_prev = calculate_states_recur(
-                tbranch,
-                tru_transitions,
-                schedule,
-                builder,
-                early_transitions
-            )?;
-            // Previous states transitioning into false branch need the conditional
-            // to be false.
-            let fal_transitions = preds.into_iter().map(|(s, g)| (s, g & !port_guard.clone())).collect();
-
-            let fal_prev = if let ir::Control::Empty(..) = **fbranch {
-                // If the false branch is empty, then all the prevs to this node will become prevs
-                // to the next node.
-                fal_transitions
-            } else {
-                calculate_states_recur(
-                    fbranch,
-                    fal_transitions,
-                    schedule,
-                    builder,
-                    early_transitions
-                )?
-            };
-
-            let prevs =
-                tru_prev.into_iter().chain(fal_prev.into_iter()).collect();
-            Ok(prevs)
+        ir::Control::If(if_stmt) => {
+            calc_if_recur(if_stmt, preds, schedule, builder, early_transitions)
         }
-        ir::Control::While(ir::While {
-            cond, port, body, ..
-        }) => {
-            if cond.is_some() {
-                return Err(Error::malformed_structure(format!("{}: Found group `{}` in with position of if. This should have compiled away.", TopDownCompileControl::name(), cond.as_ref().unwrap().borrow().name())));
-            }
+        ir::Control::While(while_stmt) => {
+            calc_while_recur(while_stmt, preds, schedule, builder, early_transitions)
+        }
+        ir::Control::Par(_) => unreachable!(),
+        ir::Control::Invoke(_) => unreachable!("`invoke` statements should have been compiled away. Run `{}` before this pass.", passes::CompileInvoke::name()),
+        ir::Control::Empty(_) => unreachable!("`empty` statements should have been compiled away. Run `{}` before this pass.", passes::CompileEmpty::name()),
+    }
+}
 
-            let port_guard: ir::Guard = Rc::clone(port).into();
+/// Essentially the same as `calculate_states_recur`, but takes an input Seq  
+/// instead of a Control. This is helpful if you have a seq that is not
+/// "wrapped" in the enum `Control`.
```

Comment:
This comment is misleading because `calculate_states_recur` just dispatches to this function to figure out what to do for a `seq`. Instead, explain what this function is doing.